File: check-rain.js
// check-rain.js
// Checks tomorrow's forecast for your city and sends a Telegram message if rain is expected.
// Run once a day (e.g. via GitHub Actions cron).

// ---- CONFIG ----
// FOR LOCAL TESTING: paste your real values directly into the 4 lines below (inside the quotes).
// Before pushing to GitHub, change them back to empty strings "" so you don't commit real keys —
// the script will then automatically use GitHub Secrets (via environment variables) instead.
const LOCAL_OPENWEATHER_API_KEY = "";
const LOCAL_CITY_NAME = "";
const LOCAL_TELEGRAM_BOT_TOKEN = "";
const LOCAL_TELEGRAM_CHAT_ID = "";

const OPENWEATHER_API_KEY =
  LOCAL_OPENWEATHER_API_KEY || process.env.OPENWEATHER_API_KEY;
const CITY_NAME = LOCAL_CITY_NAME || process.env.CITY_NAME;
const TELEGRAM_BOT_TOKEN =
  LOCAL_TELEGRAM_BOT_TOKEN || process.env.TELEGRAM_BOT_TOKEN;
const TELEGRAM_CHAT_ID = LOCAL_TELEGRAM_CHAT_ID || process.env.TELEGRAM_CHAT_ID;

if (
  !OPENWEATHER_API_KEY ||
  !CITY_NAME ||
  !TELEGRAM_BOT_TOKEN ||
  !TELEGRAM_CHAT_ID
) {
  console.error(
    "Missing config. Fill in the LOCAL_ variables at the top of the file for local testing, " +
      "or set OPENWEATHER_API_KEY / CITY_NAME / TELEGRAM_BOT_TOKEN / TELEGRAM_CHAT_ID as environment variables.",
  );
  process.exit(1);
}
// ----------------

const HOURS_TO_CHECK = 8; // look ahead this many hours for rain (roughly matches the gap between the 7am/1pm/7pm runs)

async function getForecast() {
  const url = `https://api.openweathermap.org/data/2.5/forecast?q=${encodeURIComponent(
    CITY_NAME,
  )}&appid=${OPENWEATHER_API_KEY}&units=metric`;

  const response = await fetch(url);

  if (!response.ok) {
    const errorBody = await response.text();
    throw new Error(
      `OpenWeatherMap request failed (${response.status}): ${errorBody}`,
    );
  }

  return response.json();
}

function willItRain(forecastData) {
  // The free "5 day / 3 hour forecast" endpoint returns a list of 3-hour blocks.
  // We only look at blocks within the next HOURS_TO_CHECK hours.
  const cutoffCount = Math.ceil(HOURS_TO_CHECK / 3);
  const upcomingBlocks = forecastData.list.slice(0, cutoffCount);

  const rainBlocks = upcomingBlocks.filter((block) => {
    const weatherMain = block.weather?.[0]?.main?.toLowerCase() || "";
    const pop = block.pop || 0; // "probability of precipitation", 0 to 1
    return weatherMain.includes("rain") || pop >= 0.4;
  });

  if (rainBlocks.length === 0) {
    return { rainExpected: false };
  }

  // Find the highest chance of rain among the flagged blocks, for a nicer message
  const worstBlock = rainBlocks.reduce((worst, block) =>
    block.pop > worst.pop ? block : worst,
  );

  const time = new Date(worstBlock.dt * 1000).toLocaleTimeString([], {
    hour: "2-digit",
    minute: "2-digit",
  });

  return {
    rainExpected: true,
    chance: Math.round(worstBlock.pop * 100),
    time,
    description: worstBlock.weather?.[0]?.description || "rain",
  };
}

async function sendTelegramMessage(text) {
  const url = `https://api.telegram.org/bot${TELEGRAM_BOT_TOKEN}/sendMessage`;

  const response = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      chat_id: TELEGRAM_CHAT_ID,
      text,
    }),
  });

  if (!response.ok) {
    const errorBody = await response.text();
    throw new Error(
      `Telegram request failed (${response.status}): ${errorBody}`,
    );
  }

  return response.json();
}

async function main() {
  try {
    console.log(`Checking forecast for ${CITY_NAME}...`);
    const forecastData = await getForecast();
    const result = willItRain(forecastData);

    if (result.rainExpected) {
      const message = `🌧️ Rain expected today in ${CITY_NAME}!\nAround ${result.time}, ~${result.chance}% chance (${result.description}). Bring an umbrella!`;
      console.log(message);
      await sendTelegramMessage(message);
      console.log("Telegram message sent.");
    } else {
      console.log(
        "No rain expected in the next",
        HOURS_TO_CHECK,
        "hours. No message sent.",
      );
    }
  } catch (err) {
    console.error("Error running rain check:", err.message);
    process.exit(1);
  }
}

main();

File: check_rain.py
#!/usr/bin/env python3
# check_rain.py
# Checks today's forecast for your city and sends a Telegram message if rain is expected.
# Run a few times a day (e.g. via GitHub Actions cron).

import json
import os
import sys
import urllib.error
import urllib.parse
import urllib.request
from datetime import datetime

# Windows consoles often default to a legacy codepage (e.g. cp1252) that can't
# encode the emoji in the rain message; force UTF-8 stdout/stderr so printing
# never crashes regardless of the terminal's default encoding.
for _stream in (sys.stdout, sys.stderr):
    if hasattr(_stream, "reconfigure"):
        _stream.reconfigure(encoding="utf-8")

# ---- CONFIG ----
# FOR LOCAL TESTING: paste your real values directly into the 4 lines below (inside the quotes).
# Before pushing to GitHub, change them back to empty strings "" so you don't commit real keys —
# the script will then automatically use GitHub Secrets (via environment variables) instead.
LOCAL_OPENWEATHER_API_KEY = ""
LOCAL_CITY_NAME = ""
LOCAL_TELEGRAM_BOT_TOKEN = ""
LOCAL_TELEGRAM_CHAT_ID = ""

OPENWEATHER_API_KEY = LOCAL_OPENWEATHER_API_KEY or os.environ.get("OPENWEATHER_API_KEY")
CITY_NAME = LOCAL_CITY_NAME or os.environ.get("CITY_NAME")
TELEGRAM_BOT_TOKEN = LOCAL_TELEGRAM_BOT_TOKEN or os.environ.get("TELEGRAM_BOT_TOKEN")
TELEGRAM_CHAT_ID = LOCAL_TELEGRAM_CHAT_ID or os.environ.get("TELEGRAM_CHAT_ID")

if not all([OPENWEATHER_API_KEY, CITY_NAME, TELEGRAM_BOT_TOKEN, TELEGRAM_CHAT_ID]):
    print(
        "Missing config. Fill in the LOCAL_ variables at the top of the file for local testing, "
        "or set OPENWEATHER_API_KEY / CITY_NAME / TELEGRAM_BOT_TOKEN / TELEGRAM_CHAT_ID as environment variables.",
        file=sys.stderr,
    )
    sys.exit(1)
# ----------------

HOURS_TO_CHECK = 8  # look ahead this many hours for rain (roughly matches the gap between the 7am/1pm/7pm runs)


def get_forecast():
    query = urllib.parse.urlencode(
        {"q": CITY_NAME, "appid": OPENWEATHER_API_KEY, "units": "metric"}
    )
    url = f"https://api.openweathermap.org/data/2.5/forecast?{query}"

    try:
        with urllib.request.urlopen(url) as response:
            return json.loads(response.read().decode())
    except urllib.error.HTTPError as e:
        error_body = e.read().decode()
        raise RuntimeError(f"OpenWeatherMap request failed ({e.code}): {error_body}")


def will_it_rain(forecast_data):
    # The free "5 day / 3 hour forecast" endpoint returns a list of 3-hour blocks.
    # We only look at blocks within the next HOURS_TO_CHECK hours.
    cutoff_count = -(-HOURS_TO_CHECK // 3)  # ceil division
    upcoming_blocks = forecast_data["list"][:cutoff_count]

    rain_blocks = []
    for block in upcoming_blocks:
        weather = block.get("weather") or [{}]
        weather_main = (weather[0].get("main") or "").lower()
        pop = block.get("pop", 0)  # "probability of precipitation", 0 to 1
        if "rain" in weather_main or pop >= 0.4:
            rain_blocks.append(block)

    if not rain_blocks:
        return {"rain_expected": False}

    # Find the highest chance of rain among the flagged blocks, for a nicer message
    worst_block = max(rain_blocks, key=lambda b: b.get("pop", 0))

    time_str = datetime.fromtimestamp(worst_block["dt"]).strftime("%I:%M %p")

    weather = worst_block.get("weather") or [{}]
    return {
        "rain_expected": True,
        "chance": round(worst_block.get("pop", 0) * 100),
        "time": time_str,
        "description": weather[0].get("description", "rain"),
    }


def send_telegram_message(text):
    url = f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}/sendMessage"
    data = json.dumps({"chat_id": TELEGRAM_CHAT_ID, "text": text}).encode()
    request = urllib.request.Request(
        url, data=data, headers={"Content-Type": "application/json"}, method="POST"
    )

    try:
        with urllib.request.urlopen(request) as response:
            return json.loads(response.read().decode())
    except urllib.error.HTTPError as e:
        error_body = e.read().decode()
        raise RuntimeError(f"Telegram request failed ({e.code}): {error_body}")


def main():
    try:
        print(f"Checking forecast for {CITY_NAME}...")
        forecast_data = get_forecast()
        result = will_it_rain(forecast_data)

        if result["rain_expected"]:
            message = (
                f"🌧️ Rain expected today in {CITY_NAME}!\n"
                f"Around {result['time']}, ~{result['chance']}% chance ({result['description']}). "
                "Bring an umbrella!"
            )
            print(message)
            send_telegram_message(message)
            print("Telegram message sent.")
        else:
            print(f"No rain expected in the next {HOURS_TO_CHECK} hours. No message sent.")
    except Exception as err:
        print(f"Error running rain check: {err}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
